Add optional toolbar prop to terra-navigation-side-menu

dev-cprice opened the pull request:
### Summary

Add toolbar prop to `terra-navigation-side-menu`.

[Deployment link: Component](https://terra-framework-deploye-pr-352.herokuapp.com/#/components/terra-navigation-side-menu/navigation-side-menu/navigation-side-menu)
[Deployment link: Test](https://terra-framework-deploye-pr-352.herokuapp.com/#/tests/terra-navigation-side-menu/navigation-side-menu/navigation-side-menu-default-with-toolbar)

Thanks for contributing to Terra. 
@cerner/terra

[CONTRIBUTORS.md]: ../blob/master/CONTRIBUTORS.md
[License]: ../blob/master/LICENSE

dev-cprice commented on packages/terra-navigation-side-menu/src/NavigationSideMenu.jsx:
```diff
@@ -96,20 +98,25 @@ const processMenuItems = (menuItems) => {
 };
 
-class NavigationSideMenu extends React.Component {
+const searchFilter = (searchable = '', by = '') => searchable.toLowerCase().search(by.toLowerCase()) >= 0;
```
Open to using `.includes()` instead of `.search()` here

dev-cprice commented on packages/terra-navigation-side-menu/src/NavigationSideMenu.jsx:
```diff
@@ -172,11 +189,11 @@ class NavigationSideMenu extends React.Component {
+  buildListContent(keys) {
+    return (keys && keys.length
+      ? <List className={cx(['menu-list'])}>{keys.map(key => this.buildListItem(key))}</List>
+      : <FormattedMessage id="Terra.NavigationSideMenu.noMatch" />
```
Open to show something other than "No match..." here for the empty list state.

Some other ideas:
* Show nothing (seems bad)
* Show the entire, unfiltered list

JakeLaCombe commented:
Do we have an issue logged for this?

dev-cprice commented:
@JakeLaCombe No issue logged yet. I originally intended this to be a hackathon, but had a little downtime to implement it real quick.

JakeLaCombe commented:
If we want this feature, my guess is that we should remove the rounded corners form the search. It's looking a bid strange, as everything else is full width of the container, and the corners throw them slightly off

JakeLaCombe commented:
Go ahead an log one. I'd like @mmalaker to take a look at it to see what he thinks of the feature.

tbiethman commented:
@dev-cprice This is really cool. The NavigationSideMenu is meant to be an easy starting point for teams wanting to build secondary navigation components, and I think this filtering functionality could make sense for some use cases (like the Components list). But it might not be universally needed or wanted across the consumer base, and going through the UX process to determine whether it's desired would significantly delay this feature being available.

But we do want this filtering in terra-dev-site. So here's what I recommend we do:

1. Update the NavigationSideMenu to accept a `toolbar` or similar prop. The `toolbar` content would be rendered between the header and list contents. This is how you would insert the search field into the NavigationSideMenu.
2. Duplicate the RoutingMenu component from 'terra-application-layout' and implement your changes there: https://github.com/cerner/terra-framework/blob/master/packages/terra-application-layout/src/menu/RoutingMenu.jsx. This component (DevSiteRoutingMenu or something) would be exclusively used by and live within terra-dev-site.
3. Update terra-dev-site's generation scripts to use your new component instead of the old RoutingMenu: https://github.com/cerner/terra-dev-site/blob/45b34e0045b56386c17f336b7ca62bb94fcd85d5/scripts/generate-app-config/generateRouteConfig.js#L170

This is also more involved than we want it to be, but these are changes that can be made today without any additional investigation. If it is deemed necessary for the NavigationSideMenu to include filtering in the future, we can reevaluate things.

dev-cprice commented:
@JakeLaCombe as for the removal of the rounded corners, I can only seem to get them to go away if I use an `!important`... thoughts on any other ways around this?
![image](https://user-images.githubusercontent.com/34115417/49185414-88407580-f327-11e8-9d8e-6bcc0c3035a5.png)

The [input](https://github.com/cerner/terra-core/blob/master/packages/terra-search-field/src/SearchField.module.scss#L7-L10) and [button](https://github.com/cerner/terra-core/blob/master/packages/terra-search-field/src/SearchField.module.scss#L26-L29) are both themed to have `3px` rounding by default.

JakeLaCombe commented:
Given that we have it just for dev site and not production, I would be okay with leaving the rounded corners for now.

dev-cprice commented:
Reverted `terra-navigation-side-menu` back to previous behavior, but with toolbar prop added. Added duplicate of `terra-application-layout`'s `RoutingMenu` with the filtering of the menuItems. Opening a PR on `terra-dev-site` to use the new `DevSiteRoutingConfig` in [generateRouteConfig](https://github.com/cerner/terra-dev-site/blob/master/scripts/generate-app-config/generateRouteConfig.js#L6). Will comment here.

mjhenkes commented:
Could you move `DevSiteRoutingMenu.jsx` to [here](https://github.com/cerner/terra-dev-site/tree/master/src/app/components) in the dev site project?

neilpfeiffer commented:
I see the value of adding a search/filter utility, but hate to be the one to rain on the parade with the current approach.  Our current dev-site from an information architecture/design standpoint, structured differently than Storybook or other applications or websites like [Semantic UI](https://react.semantic-ui.com/) that is only a left nav.  Our dev-site currently is based on the application framework layout and is structured more like a traditional content-driven website with primary nav across the top and so a filter within a left navigation is out of place and not a great ui pattern, in favor of a traditional site/application search in the top header bar like [Material UI](https://material-ui.com/api/button/) and [Ant UI](https://ant.design/components/button/).

So as to not stop progress and mature the design a bit, my idea is to see if we can make it a dev-site only exclusive that has opt-in configuration, and marked as beta feature  and that could change in location/implementation in the future. I don't want roll it out to terra-ui.com until more of a cohesive design and approach has been worked through.

bjankord commented:
@neilpfeiffer We are in the process of moving the code over to the dev site. See @mjhenkes's comment [here](https://github.com/cerner/terra-framework/pull/352#issuecomment-442841936). I'm ok with making it opt-in configuration in dev site, and marked as beta feature that could change in location/implementation in the future but I want to add this to terra-ui.com as-is. We can iterate on it, but I know if we wait for the turn around time it will take to have more of a cohesive design and approach that has been worked through and widely introducing it as standard feature means that this is likely not going to be rolled out to terra-ui.com until a much much later date. This is a highly desired feature by our users, I'd rather get a mvp out the door than wait for a fully Cerner design standard approved implementation to land.

neilpfeiffer commented:
Lets regroup on a plan offline, I definitely see the value and want to help keeping thing moving and mvp, but needs to be inline with minimum best practices even in an initial state.

bjankord commented:
Talked offline:

Want to update so:
* It only shows up if there are 10 or more

Future enhancements:
* Improve design on input
* Improve no status visual

dev-cprice commented:
@bjankord @neilpfeiffer comments addressed here: https://github.com/cerner/terra-dev-site/pull/127

bjankord commented:
@dev-cprice Looks like jest snapshots need to be updated.

yuderekyu commented on packages/terra-navigation-side-menu/src/NavigationSideMenu.jsx:
```diff
@@ -173,10 +177,10 @@ class NavigationSideMenu extends React.Component {
-      return <List className={cx(['menu-list'])}>{currentItem.childKeys.map(key => this.buildListItem(key))}</List>;
-    }
-    return null;
+    return (currentItem && currentItem.childKeys && currentItem.childKeys.length
```
not a big deal, but to me, this logic is harder to understand when it is a ternary statement

yuderekyu commented on packages/terra-navigation-side-menu/src/terra-dev-site/test/navigation-side-menu/NavigationSideMenuDefaultWithToolbar.test.jsx:
```diff
@@ -0,0 +1,77 @@
+          id="test-menu"
+          menuItems={[
+            { key: 'menu', text: 'Menu', childKeys: ['submenu1', 'submenu2', 'submenu3', 'submenu4'] },
+            {
```
nit, want to make this object a single line like the others?

dev-cprice replied:
this was just copy-pasted.. probably from auto lint change

emilyrohrbough commented on dev-site-config/site.config.js:
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  filterSideMenu: true,
```
Should we wait to update this config for when terra-dev-site is released?

bjankord replied:
I think this should be fine, I don't see it having any impact if the config doesn't do anything in terra-dev-site yet.

emilyrohrbough commented:
Looks like theres a bad screenshots thats been included: 

![screen shot 2018-12-03 at 10 48 01 am](https://user-images.githubusercontent.com/14099737/49388117-f77fe600-f6e8-11e8-8ae7-d6dd0876fbdc.png)

emilyrohrbough commented on packages/terra-navigation-side-menu/CHANGELOG.md:
```diff
@@ -6,6 +6,8 @@ Unreleased
 ### Changed
 * Wdio tests changed to use themeCombinationOfCustomProperties
 
+* Add optional toolbar prop that displays below the side menu action header
```
Terra follows [these format guidelines](https://keepachangelog.com/en/1.0.0/) for writing change log entries. Lets add the heading `### Added` for this entry.

emilyrohrbough commented on packages/terra-application-layout/tests/jest/__snapshots__/DevSiteRoutingMenu.test.jsx.snap:
```diff
@@ -0,0 +1,52 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
```
Looks like this should have been removed from this PR when you made changes to migrate the DevSiteRouting Menu to the terra-dev-site repo

emilyrohrbough commented on packages/terra-application-layout/package.json:
```diff
@@ -47,7 +47,8 @@
     "terra-navigation-side-menu": "^1.16.0",
-    "terra-popup": "^4.18.1"
+    "terra-popup": "^4.18.1",
+    "terra-search-field": "^2.36.1"
```
I think we can drop the `terra-search-field` dependency now

bjankord commented:
@dev-cprice Looks like there are some lint errors that need to be addressed.

dev-cprice commented:
@bjankord lint and build fixed 🙌 :shipit:
